ensips/ccip-write/ccip-write.md

```diff
@@ -0,0 +1,342 @@
+3. **L2 Exits**: For L2 implementations, secure exit mechanisms MUST be provided to allow users to withdraw their domain ownership and data back to L1 in case of L2 failure or other emergencies.
+4. **Data Availability Challenges**: L2 implementations MUST have robust mechanisms to ensure that crucial ENS data remains available and verifiable on L1, even in scenarios where the L2 network faces challenges or downtime.
+
+Further security analysis and auditing are RECOMMENDED before deploying this system in a production environment, with special attention given to the unique security considerations of both database and L2 implementations.
+
+## Appendix
+
+### Appendix 1: Offchain Storage Implementation with EIP-5559 changes
+
+Although EIP-5559 is still under community discussion, it would be a huge improvement if the parameters were encoded as native bytes format, changing the `parameters` struct to the encoded calldata in `bytes` format and removing the type cast library from the contract. The `StorageHandledByOffChainDatabase` revert implementation as specified by EIP-5559:
+
+```solidity
+// current IWriteDeferral implementation
+struct messageData {
+  bytes4 functionSelector;
+  address sender;
+  parameter[] parameters;
+  uint256 expirationTimestamp;
+}
+
+// proposed IWriteDeferral implementation
+struct messageData {
+  bytes callData; // encoded version of function signature and its arguments
+  address sender;
+  uint256 expirationTimestamp;
+}
+```
+
+```solidity
+/**
+ * @notice Builds a StorageHandledByOffChainDatabase error.
+ * @param params The offChainDatabaseParamters used to build the corresponding mutation action.
+ */
+function _offChainStorage(IWriteDeferral.parameter[] memory params)
+    private
+    view
+{
+    revert StorageHandledByOffChainDatabase(
+        IWriteDeferral.domainData({
+            name: _WRITE_DEFERRAL_DOMAIN_NAME,
+            version: _WRITE_DEFERRAL_DOMAIN_VERSION,
+            chainId: _CHAIN_ID,
+            verifyingContract: address(this)
+        }),
+        gatewayUrl,
+        IWriteDeferral.messageData({
+            callData: msg.data,
+            sender: msg.sender,
+            expirationTimestamp: block.timestamp
+                + gatewayDatabaseTimeoutDuration
+        })
+    );
+
+```
+
+### Appendix 2: CCIP-Store
+
+Even though the recently proposed offchain writing strategy specified in [EIP-7700](https://eips.ethereum.org/EIPS/eip-7700#database-router-storageroutedtodatabase) handles the signature step differently, the current `register` signature is still a fully compliant way of registering a new domain.
+
+### Appendix 3: IOffchainResolver interface
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+/**
+ * @title IOffchainResolver Interface
+ * @dev Interface for a resolver that supports text and address records management, with additional functions for registration, ownership transfer, and multicall.
+ */
+interface IOffchainResolver {
```

Comment:
Apply changes from above here. 

Also add the commit interface and the explanation for it